include/BaselineGenSelection.h

```diff
@@ -5,11 +5,297 @@
+                        GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass);
+  }
+
+  if (!decays_to_hadrons && !decays_to_leptons)
+  {
+    throw analysis::exception("Encountered neither leptonic nor hadronic vector boson decay");
+  }
+
+  return res;
+}
+
+HVVCand GetGenHVVCandidate(int evt, std::vector<reco_tau::gen_truth::GenLepton> const& gen_leptons,
+                           const RVecI& GenPart_pdgId, const RVecVecI& GenPart_daughters, const RVecI& GenPart_statusFlags,
+                           const RVecF& GenPart_pt, const RVecF& GenPart_eta, const RVecF& GenPart_phi, const RVecF& GenPart_mass,
+                           bool throw_error_if_not_found)
+{
+  try
+  {
+    int sz = GenPart_pdgId.size();
+    std::set<int> HVV_indices;
+    for (int i = 0; i < sz; ++i)
+    {
+      const GenStatusFlags status(GenPart_statusFlags.at(i));
+      bool is_higgs = GenPart_pdgId[i] == PdG::Higgs();
+      if (!(is_higgs && status.isLastCopy()))
+      {
+        continue;
+      }
+      auto const& daughters = GenPart_daughters.at(i);
+      auto IsV = [&](int idx){ return GenPart_pdgId[idx] == PdG::Wplus() || GenPart_pdgId[idx] == PdG::Wminus() || GenPart_pdgId[idx] == PdG::Z(); };
+      int n_V_from_H = std::count_if(daughters.begin(), daughters.end(), IsV);
+      if (n_V_from_H == 2)
+      {
+        HVV_indices.insert(i);
+      }
+    }
+
+    if(HVV_indices.empty())
+    {
+      throw analysis::exception("H->VV not found.");
+    }
+    if(HVV_indices.size() != 1)
+    {
+      throw analysis::exception("Multiple H->VV candidates.");
+    }
+
+    int const HVV_index = *HVV_indices.begin();
+    HVVCand HVV_cand;
+
+    HVV_cand.cand_p4 = GetP4(GenPart_pt, GenPart_eta, GenPart_phi, GenPart_mass, HVV_index);
+
+    RVecI V_from_H = GenPart_daughters.at(HVV_index);
+    int v1_idx = V_from_H.at(0);
+    int v2_idx = V_from_H.at(1);
+
+    v1_idx = GetLastCopy(v1_idx, GenPart_pdgId, GenPart_statusFlags, GenPart_daughters);
+    v2_idx = GetLastCopy(v2_idx, GenPart_pdgId, GenPart_statusFlags, GenPart_daughters);
+
+    RVecI daughters_v1 = GenPart_daughters.at(v1_idx);
+    bool v1_to_hadrons = std::all_of(daughters_v1.begin(), daughters_v1.end(),
+                                      [&](int idx){ return std::abs(GenPart_pdgId[idx]) <= PdG::b(); });
+    bool v1_to_leptons = std::all_of(daughters_v1.begin(), daughters_v1.end(),
+                                      [&](int idx){ return PdG::isNeutrino(GenPart_pdgId[idx]) || PdG::isLepton(GenPart_pdgId[idx]); });
+
+    RVecI daughters_v2 = GenPart_daughters.at(v2_idx);
+    bool v2_to_hadrons = std::all_of(daughters_v2.begin(), daughters_v2.end(),
+                                      [&](int idx){ return std::abs(GenPart_pdgId[idx]) <= PdG::b(); });
+    bool v2_to_leptons = std::all_of(daughters_v2.begin(), daughters_v2.end(),
+                                      [&](int idx){ return PdG::isNeutrino(GenPart_pdgId[idx]) || PdG::isLepton(GenPart_pdgId[idx]); });
+
```

Comment:
do it in a loop

Reply:
Access to daughters of both vector bosons simultaneously is necessary check that the H->VV decays belongs to double lepton channel or single lepton channel. It is impossible to implement it in a loop.